Refactor yum and dnf, add feature parity

maxamillion opened the pull request:
Signed-off-by: Adam Miller <admiller@redhat.com>

##### SUMMARY
Bringing the `yum` and `dnf` modules into feature parity and move both to an object oriented model to reduce redundant information passing between functions (hopefully making them both more maintainable). 

A main goal of this is for the `dnf` module to have feature parity with the `yum` module but also be able to extend beyond that of the `yum` module as upstream `dnf` continues to add more features, this is in service of allowing the `yum` Ansible module to remain compatible with [YUM4](https://wiki.centos.org/SpecialInterestGroup/ConfigManagementSIG/YUM4) which actually has a dnf backend. (The compatibility will be handled with a yum action plugin in a separate pull request, I wanted to try and limit the scope of each PR)

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request


##### COMPONENT NAME
<!--- Name of the module, plugin, task or feature -->
yum
dnf

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.7.0.dev0 (refactor/yum_dnf_parity 29ecf9b5b7) last updated 2018/08/02 16:44:51 (GMT -500)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/admiller/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/admiller/src/dev/ansible/lib/ansible
  executable location = /home/admiller/.virtualenvs/ansible/bin/ansible
  python version = 2.7.15 (default, May 16 2018, 17:50:09) [GCC 8.1.1 20180502 (Red Hat 8.1.1-1)]
```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

maxamillion committed:
Refactor yum and dnf, add feature parity

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion committed:
remove unnecessary module_utils, move the classes into the module code

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion committed:
remove yum -> yum4, out of scope

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion committed:
use ABCMeta

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion committed:
re-arrange run() caller vs callee

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion committed:
make sanity checks happy

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion committed:
fix yum unit tests

Signed-off-by: Adam Miller <admiller@redhat.com>

mkrizek commented on lib/ansible/module_utils/yumdnf.py:
```diff
@@ -0,0 +1,122 @@
+    def run(self):
+        raise NotImplementedError
+
+    def fetch_rpm_from_url(self, spec):
```
Hopefully we'll move this into utils soon https://github.com/ansible/ansible/pull/19172

mkrizek commented on lib/ansible/module_utils/yumdnf.py:
```diff
@@ -0,0 +1,122 @@
+    """
+    Abstract class that handles the population of instance variables that should
+    be identical between both YUM and DNF modules because of the feature parity
+    and shared argement spec
```
s/argement/argument :)

mkrizek commented on test/integration/targets/dnf/tasks/dnf.yml:
```diff
@@ -241,6 +274,8 @@
     state: present
   register: dnf_result
 
+- debug: var=dnf_result
```
I wonder if we should put something like `verbosity: 1` for debug prints?

maxamillion replied:
@mkrizek that was actually an accidental left-over for when I was working on this, it's not needed for the tests

mkrizek commented on lib/ansible/modules/packaging/os/yum.py:
```diff
@@ -146,6 +150,22 @@
         The disabled plugins will not persist beyond the transaction.
     required: false
     version_added: "2.5"
+  releasever:
```
It makes me a little nervous that we're not adding tests for new features.

maxamillion replied:
Yeah, I should really add more tests. The problem is that I'm not sure how best to test this just yet. Let me think about it a little and try to get something added that has a reasonable assertion.

maxamillion replied:
However, we don't have test coverage of that functionality for the `yum` module and I brought over roughly a dozen new options to `dnf` from `yum`.... but I do agree we still need test coverage. I'll look at this when I'm back.

mkrizek reviewed (commented):
I love that we're getting rid of the long method signatures by extracting the arguments into instance vars +1000. It's huge diff to review, since all tests pass I think we should merge ASAP if we want this in `2.7` so we can have that tested properly in `devel`.

maxamillion committed:
remove unecessary debug statements, fix typo

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion committed:
fix licensing and attribution in yumdnf module_util

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion committed:
include fix from PR 40737

original commit 5cbda9658ab13d45a210c0581458327a4d975bcc
original Author: Strahinja Kustudic <kustodian@gmail.com>

yum will fail on 'No space left on device', fixes #32791 (#40737)

During the installing of packages if yum runs out of free disk space,
some post install scripts could fail (like e.g. when the kernel
package generates initramfs), but yum would still exit with a status
0.  This is bad, especially for the kernel package, because it makes
it unable to boot.  Because the yum module is usually used for
automation, which means the users cannot read every message yum
prints, it's better that the yum module fails if it detects that
there is no free space on the disk.

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion committed:
Revert "fix licensing and attribution in yumdnf module_util"

This reverts commit 59e11de5a2a6efa17ac3f0076bb162348c02e1bd.

maxamillion committed:
move fetch_rpm_from_url out of yumdnf module_util

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion committed:
fix the move of fetch_rpm_from_url

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion commented:
As per @abadger on irc, as 2.7 release manager, this module refactor was approved (mentioned here for posterity)

maxamillion commented:
rebuild_merge

sej7278 commented:
This doesn't seem to work for me, at least not in the same way as ```package-cleanup --leaves``` or ```yum autoremove``` do.

plays tried:

```
- name: prune packages
  yum: name=* autoremove=yes

- name: remove some packages
  yum:
    name:
      - kernel-headers
      - blah
    state: absent
    autoremove: yes
```

doesn't remove anything other than what i've asked it to remove, but then if i run ```yum autoremove``` it suggests this lot:

```
# yum autoremove
Loaded plugins: ulninfo
Resolving Dependencies
--> Running transaction check
---> Package fxload.x86_64 0:2002_04_11-16.el7 will be erased
---> Package libndp.x86_64 0:1.2-7.el7 will be erased
---> Package lsscsi.x86_64 0:0.27-6.el7 will be erased
---> Package mariadb-libs.x86_64 1:5.5.60-1.el7_5 will be erased
---> Package teamd.x86_64 0:1.27-5.el7 will be erased
--> Finished Dependency Resolution
--> Finding unneeded leftover dependencies
---> Marking libdaemon to be removed - no longer needed by teamd
---> Marking libnl3 to be removed - no longer needed by teamd
---> Marking libteam to be removed - no longer needed by teamd
---> Marking libnl3-cli to be removed - no longer needed by teamd
Found and removing 4 unneeded dependencies
--> Running transaction check
---> Package libdaemon.x86_64 0:0.14-7.el7 will be erased
---> Package libnl3.x86_64 0:3.2.28-4.el7 will be erased
---> Package libnl3-cli.x86_64 0:3.2.28-4.el7 will be erased
---> Package libteam.x86_64 0:1.27-5.el7 will be erased
--> Finished Dependency Resolution
```

ansible 2.7.1 and oel 7.6 (yum 3.4.3, rpm 4.11)

alex-mohemian commented:
I have the same issue as @sej7278 !

sej7278 commented:
@alex-mohemian yes that's what i've been using but not too happy with shelling out (and hiding the warning about doing so!) can't find where it says its not possible though.

maxamillion commented:
Yes, `autoremove` was broken by mistake. The issue should be fixed once Ansible 2.7.2 is released https://github.com/ansible/ansible/pull/47902

alex-mohemian commented:
@maxamillion thank you. Are you sure the behavior is fixed and that 

```
- name: cleanup
  yum: autoremove=yes
```

will do what `sudo yum autoremove -y` does? From https://github.com/ansible/ansible/pull/47902 I don't see that...

sej7278 commented:
just tried rhel8 beta1 and ansible is totally broken (not least by no python interpreters!)

but these don't work at all:

```
- name: prune packages
  dnf:
    autoremove: yes

- name: prune packages
  dnf: autoremove=yes

fatal: [vbrhel8]: FAILED! => {"changed": false, "msg": "Autoremove should be used alone or with state=absent", "results": []}

```

maxamillion commented:
@sej7278 did you try with current devel or wtih the `stable-2.7` branch from source? the fix is still unreleased

maxamillion commented:
@alex-mohemian yes, the issue was that the value was being set incorrectly for `self.autoremove` and `self.state` that were being checked to then invoke `yum -y autoremove` so the change only sets those correctly, the conditional to set the command call on the backend didn't need modification.

sej7278 commented:
i'm using ansible debian package 2.7.1+dfsg-2 which i assume has neither yum not dnf autoremove fixes (its from about 28th october)
